Bug Tracking (#4)
forwards/rdevguide: 11-bug_tracking.Rmd

2. Issues with $\textit{seemingly}$ identical numbers not being equal (especially floating point numbers) are usually not bugs. 

3. If $\textsf{R}$ is running slower than expected, then also it may not be a bug. Ask someone else to review your code in such a case.

4. If some function is working, but it is not defined in the best generalised way, then consult some to look over your code. This may perhaps not be a bug, instead it might be an alternative way of writing the function.

## Checking if a bug is already reported

The first step before filing a bug report is to see whether the problem has already been reported. Checking if the bug is reported will:

1. Save time for you and the developers

2. Help you see if the bug is already fixed for the next release

3. Lead you to learn what needs to be done to fix it

4. Determine if any additional information is needed

The sections that follow discuss where to check whether a bug is already reported.

## Levels of contributing to bug / What do you do when you find a bug?

1. [Report](#ReportBug) the bug (if it is not alreasy reported).

2. [Test](provide hyperlink to the section/chapter which discusses this) the bug.

> 3. [Fix](hyperlink to the chapter on the lifecycle of a patch) the bug: Fixing a bug might require relatively more time. You may start a conversation about it either on BugZilla (link here) or on the GitHub mirror (link here). This would require engagement with $\textsf{R}$ Core team.

Author: Fixing bugs will be discussed in the chapter on lifecycle of a patch. How to test bugs will probably be a part of the next chapter on reviewing bugs. Let's see how those chapters shape up!

> 3. [Fix](hyperlink to the chapter on the lifecycle of a patch) the bug: Fixing a bug might require relatively more time. You may start a conversation about it either on BugZilla (link here) or on the GitHub mirror (link here). This would require engagement with $\textsf{R}$ Core team.

Reviewer: Ok, let's see. I hope that the different chapters can be merged soon, it will help to have a more general read of the book and see if the sections match well. Is there any plan on when to merge #3 or this PR?

> 3. [Fix](hyperlink to the chapter on the lifecycle of a patch) the bug: Fixing a bug might require relatively more time. You may start a conversation about it either on BugZilla (link here) or on the GitHub mirror (link here). This would require engagement with $\textsf{R}$ Core team.

Author: I hope it is merged soon.

## What are some places where you may find a bug?

You may find a bug in:

1. The [documentation of this guide](#GuideDocBug).

2. In the [$\textsf{R}$ Core-supported packages and/or their documentations](#RCorePkgBug).

3. In the [$\textsf{R}$ language](#RLangBug).

4. In [packages and/or their documentations which are not supported by the $\textsf{R}$ Core](#nonRCorePkgBug).

## How to report a bug? {#ReportBug}

Once you are confirmed that there is a bug, you need to submit a report so that it gets fixed.

### Bug in the documentation of this guide {#GuideDocBug}

1. If there is a bug in the documentation of this guide, then please report it as an issue at https://github.com/forwards/rdevguide/issues. 

> 2. Please go through the $\textit{closed}$ issues to check if the bug is already reported and fixed. Otherwise, please check in the $\textit{open}$ issues to see if it is reported. If it is not recorded at either of these places, then please submit a new issue.

Author: In the introduction or as the last of the four sub-sub-sections of "Bug in ..."? What would be less confusing to readers? What do you suggest?

> 2. Please go through the $\textit{closed}$ issues to check if the bug is already reported and fixed. Otherwise, please check in the $\textit{open}$ issues to see if it is reported. If it is not recorded at either of these places, then please submit a new issue.

Reviewer: I think that improvements on the book itself should be addressed on the introduction of the book (so on index.Rmd itself). It will be a great way to encourage more participation of people on the book.

> 2. Please go through the $\textit{closed}$ issues to check if the bug is already reported and fixed. Otherwise, please check in the $\textit{open}$ issues to see if it is reported. If it is not recorded at either of these places, then please submit a new issue.

Author: Ok. My plan is to write the introduction to the book chapter after some chapters are completed. Yes, it sounds good to include the improvements to the book and reporting bugs in this documentation, in that chapter.

### Bug in the $\textsf{R}$ Core-supported packages and/or their documentations {#RCorePkgBug}

1. Packages that are supported by the $\textsf{R}$ Core are labelled with `Maintainer: R Core Team <R-core@r-project.org>`. You can check this label either by running `maintainer("package_name")` or by looking at the `Maintainer` field in the output when the command `packageDescription("package_name")` is run in $\textsf{R}$.

> 2. The bug report for $\textsf{R}$ Core-supported packages and/or their documentations can be submitted either to $\textsf{R}$'s [Bugzilla](https://bugs.r-project.org/bugzilla/) or as an issue to the [GitHub Mirror of R svn server](https://github.com/r-devel/r-svn/issues).

Reviewer: I had the impression that the repo r-devel/r-svn was not officially supported as a place to report bugs. It is only a community effort to make it easier to submit patches. Has this changed? Is there any reference to support this?

Author: Yes, it is still a community effort to make submitting patches and/or issues easier.

Reviewer: Then I would change the wording to something like:

> 2. The bug report for $\textsf{R}$ Core-supported packages and/or their documentations must be submitted to $\textsf{R}$'s [Bugzilla](https://bugs.r-project.org/bugzilla/).

Reviewer 2: I agree with @llrs, I definitely do not feel that this guide should suggest reporting, tracking, or commenting on bugs in R via the github mirror. My understanding is the same as his.

4. Please ensure whether the bug is already fixed (in the upcoming changes in R) or reported (search for it from those already reported on Bugzilla, either on [search existing bug reports](https://bugs.r-project.org/bugzilla/query.cgi), may use the [advanced search](https://bugs.r-project.org/bugzilla/query.cgi?format=advanced) option here,  or [show open bugs new-to-old](https://bugs.r-project.org/bugzilla/buglist.cgi?bug_file_loc_type=allwordssubstr&bug_status=NEW&bug_status=ASSIGNED&bug_status=CONFIRMED&bug_status=REOPENED&bug_status=UNCONFIRMED&bugidtype=include&chfieldto=Now&cmdtype=doit&emailassigned_to1=1&emailassigned_to2=1&emailcc2=1&emailreporter2=1&emailtype1=substring&emailtype2=substring&field0-0-0=noop&long_desc_type=substring&order=bugs.delta_ts%20desc&query_format=advanced&short_desc_type=allwordssubstr&type0-0-0=noop)). If you are submitting the report as an issue on the GitHub Mirror, then please check the currently $\textit{open}$ and $\textit{closed}$ issues first, to find if it is already reported or resolved. Do not submit a report in such a case. 

### Bug in the $\textsf{R}$ Language {#RLangBug}

> 1. If there is a bug in the $\textsf{R}$ language itself, then you need to report using the same procedure and precautions as explained [above](#RCorePkgBug).

Reviewer: If there isn't a second paragraph, I don't think it makes sense to enumerate them.  Also I would merge the #RCorePkgBug and the #RLangBug section in one. To distinguish between R-core and community supported packages.

### Bug in the non $\textsf{R}$ Core-supported packages and/or their documentations {#nonRCorePkgBug}

1. For packages that are not maintained by the $\textsf{R}$ Core, the bug reports are usually submitted to their maintainer(s). You can find who is the package maintainer by running either of the commands `maintainer("package_name")` or `packageDescription("package_name")`.

> 2. In $\textsf{R}$ the running function `bug.report(package = "package_name")` shall direct you to either the GitHub issue tracker of the package, or to the bug tracking web page, or towards composing an e-mail to the package maintainer. This function `bug.report` is disabled in $\textbf{RStudio}$.

Reviewer: Just tested this. Using `bug.report("MASS")` on RStudio doesn't work as expected on RStudio. However if one uses `utils::bug.report(package = "MASS")` it works.

Reviewer 3: Interesting! Also I would not enumerate these paragraphs, the `bug.report()` function is a helper to follow the process I described above, so it is the same point, not a sequence of points.

> 3. Please ensure that your bug is not already reported or fixed before reporting it in any of the ways suggested above.

Reviewer: If this paragraph is for #nonRCorePkgBug, #RCorePkgBug and the #RLangBug section it should be on a different section or out of this. Perhaps it could be expanded to explain how to ensure that the bug is not already reported or fixed and what to do in those occasions that it is already solved/reported. Some suggestions, search error or function name not the whole code.

## Good practices in reporting bugs / Expectations of a (good) bug report

If you follow the practices given below, you will come up with a good bug report which might make it easier for the Maintainer(s) to fix the bug.

1. Include a minimal reproducible example of the bug in your report. The maintainer should be able to quickly reproduce the bug on using the minimal example that you provide.

> 

Reviewer: Not sure how agnostic this book should be but perhaps mention tools like reprex, or other documentation about how to make a minimal reproducible example ? The great [SO question](https://stackoverflow.com/q/5963269/2886003)

Reviewer 3: Linking to the SO question is a good idea - as it is a community wiki post, that is quite neutral and gives generally useful advice. We might want to emphasise specific points - e.g. use inbuilt data sets as far as possible. Again, something to get feedback from R Core on.

> 2. Mention the software architecture on which the bug occurred. 

Reviewer: This section seems only for #RCorePkgBug and the #RLangBug, perhaps it would be better to add a note at the beginning of the section? (Not that for some other packages this information is not relevant but as the links just point to r-project better to be explicit)

In addition to the above, here are the [bug writing guidelines](https://bugs.r-project.org/bugzilla/page.cgi?id=bug-writing.html) on Bugzilla. The [bug reporting](https://www.r-project.org/bugs.html#writing-a-good-bug-report) documentation in R also the discusses practices to write a good bug report.

Once you have successfully reported a bug, you will likely receive an update each time an action is taken on the bug. On Bugzilla, the report may be given one of the following status: New, Assigned, Confirmed, Reopened, Unconfirmed.

> 

Reviewer: When I first submitted a bug I didn't know the Severity to give to a bug or the Component to use, perhaps some paragraphs about each element of a bug report will make it easier for people to understand how to make a good bug report on Bugzilla

Reviewer 3: Yes. This would be good to discuss e.g. at the RCWG and get input from R core members.

Reviewer 3: Yes. This would be good to discuss e.g. at the RCWG and get input from R core members.

## Disagreement with a resolution on the bug tracker

As humans, there might be differences of opinions from time to time. What needs to be considered here is that, we are respectful of the fact that care, thought, and volunteer time has gone into the resolution of the issue or bug. 

If you take some time, then on reflection, the resolution steps may seem more reasonable than you initially thought. If you still feel that the resolution is incorrect, then raise a thoughtful question to the person who resolved it. If the issue was carefully thought about in the first place then is less likely to win any conversion of thought.

> As a reminder, issues closed by a core developer on the GitHub Mirror have already been carefully considered. Please do not reopen a closed issue.

Reviewer: Perhaps mention that even if an issue or bug report is closed one can still comment on it.

Reviewer 3: This should be Bugzilla vs GitHub mirror here.

Reviewer 2: I have not read the whole document yet so it may be mentioned elsewhere, but every comment generates an email to every R-core member (or at least ones that don't have notifications disabled), and so care should be taken when commenting on issues in a couple of different cases, which include a) issues that have already been considered and closed, and b) commenting in pure agreement without adding anything beyond that to a discussion (+1 type posts which are perfectly acceptable in other contexts).

```{r, results='hide'}
##  *** caught segfault ***
## address (nil), cause 'memory not mapped'
```

2. If the code does not do what the documentation says it should, then either the code or the documentation is wrong. Report either of which needs to be fixed.

$\textbf{Note}$:
When you are in doubt that there is a bug:

1. Make sure whether the bug appears in a clean session of $\textsf{R}$. Many a times, there are variables/commands/functions stored in history which might causes issues. Hence, check in a clean session. To do so, launch $\textsf{R}$ from the command line with the `--vanilla` option. 

2. At times the code that is written is very complicated, has numerous package and file dependencies, has many function calls, etc.. In such scenarios it is quite common that the code throws an error and you are not able to solve it. You may tend to think that there is a bug that needs to be reported. Before doing so, try to produce a minimum working example of the code for the section where the error occurred. Add only those package and files which are required by that section, and see if the error appears still. Using this approach shall solve most of the errors.

> ## What may not be a bug?

Reviewer 4: ```suggestion
## What condition might not be a bug?

```

1. In case the code is doing something unexpected, it may not necessarily be a bug. Carefully review the documentation of the function being called, and check whether the behaviour being exhibited on calling this function is the same as it was designed to do.

2. Issues with $\textit{seemingly}$ identical numbers not being equal (especially floating point numbers) are usually not bugs. 

3. If $\textsf{R}$ is running slower than expected, then also it may not be a bug. Ask someone else to review your code in such a case.

> 4. If some function is working, but it is not defined in the best generalised way, then consult some to look over your code. This may perhaps not be a bug, instead it might be an alternative way of writing the function.

Reviewer 4: ```suggestion
4. If some function is working, but it is not defined in the best generalised way, then consult someone to look over your code. This may perhaps not be a bug; instead, it might be an alternative way of writing the function.

```

2. Help you see if the bug is already fixed for the next release

3. Lead you to learn what needs to be done to fix it

4. Determine if any additional information is needed

The sections that follow discuss where to check whether a bug is already reported.

## Levels of contributing to bug / What do you do when you find a bug?

1. [Report](#ReportBug) the bug (if it is not already reported).

2. [Test](provide hyperlink to the section/chapter which discusses this) the bug.

3. [Fix](hyperlink to the chapter on the lifecycle of a patch) the bug: Fixing a bug might require relatively more time. You may start a conversation about it either on BugZilla (link here) or on the GitHub mirror (link here). This would require engagement with $\textsf{R}$ Core team.

## What are some places where you may find a bug?

You may find a bug in:

1. The [documentation of this guide](#GuideDocBug).

2. In the [$\textsf{R}$ Core-supported packages and/or their documentations](#RCorePkgBug).

3. In the [$\textsf{R}$ language](#RLangBug).

4. In [packages and/or their documentations which are not supported by the $\textsf{R}$ Core](#nonRCorePkgBug).

## How to report a bug? {#ReportBug}

> Once you are confirmed that there is a bug, you need to submit a report so that it gets fixed.

Reviewer 4: ```suggestion
Once you confirm a bug exists, you need to submit a bug report so that it gets fixed.

```

### Bug in the documentation of this guide {#GuideDocBug}

1. If there is a bug in the documentation of this guide, then please report it as an issue at https://github.com/forwards/rdevguide/issues. 

> 2. Please go through the $\textit{closed}$ issues to check if the bug is already reported and fixed. Otherwise, please check in the $\textit{open}$ issues to see if it is reported. If it is not recorded at either of these places, then please submit a new issue.

Reviewer 4: ```suggestion
2. Prior to reporting the issue, please go through the $\textit{closed}$ issues to check if the bug is already reported and fixed. Otherwise, please check in the $\textit{open}$ issues to see if it is reported. If it is not found at either of these places, then please submit a new issue.

```

### Bug in the $\textsf{R}$ Core-supported packages and/or their documentations {#RCorePkgBug}

1. Packages that are supported by the $\textsf{R}$ Core are labelled with `Maintainer: R Core Team <R-core@r-project.org>`. You can check this label either by running `maintainer("package_name")` or by looking at the `Maintainer` field in the output when the command `packageDescription("package_name")` is run in $\textsf{R}$.

> 2. The bug report for $\textsf{R}$ Core-supported packages and/or their documentations can be submitted either to $\textsf{R}$'s [Bugzilla](https://bugs.r-project.org/bugzilla/) or as an issue to the [GitHub Mirror of R svn server](https://github.com/r-devel/r-svn/issues).

Reviewer 4: ```suggestion
2. The bug report for $\textsf{R}$ Core-supported packages and/or their documentations can be submitted either to $\textsf{R}$'s [Bugzilla](https://bugs.r-project.org/bugzilla/). In the future, we hope to have an option to report an issue to the [GitHub Mirror of R svn server](https://github.com/r-devel/r-svn/issues).

```

Author: This looks like a good way to mention the GitHub Mirror of R svn server. Let's see what more suggestions come for this point.

### Bug in the $\textsf{R}$ Language {#RLangBug}

1. If there is a bug in the $\textsf{R}$ language itself, then you need to report using the same procedure and precautions as explained [above](#RCorePkgBug).

### Bug in the non $\textsf{R}$ Core-supported packages and/or their documentations {#nonRCorePkgBug}

> 1. For packages that are not maintained by the $\textsf{R}$ Core, the bug reports are usually submitted to their maintainer(s). You can find who is the package maintainer by running either of the commands `maintainer("package_name")` or `packageDescription("package_name")`.

Reviewer 3: This is probably out of date now - it is less common to have to email the maintainer these days. The best practice is to provide a url where bug reports can be submitted (an issues tracker on GitHub/GitLab/R-Forge). So I would say to look at the package DESCRIPTION file first (e.g. using `packageDescription()`) to check if a url is provided in the `BugReports` field. If that is not available. Then contact the maintainer.

2. In $\textsf{R}$ the running function `bug.report(package = "package_name")` shall direct you to either the GitHub issue tracker of the package, or to the bug tracking web page, or towards composing an e-mail to the package maintainer. This function `bug.report` is disabled in $\textbf{RStudio}$.

3. Please ensure that your bug is not already reported or fixed before reporting it in any of the ways suggested above.

## Good practices in reporting bugs / Expectations of a (good) bug report

If you follow the practices given below, you will come up with a good bug report which might make it easier for the Maintainer(s) to fix the bug.

1. Include a minimal reproducible example of the bug in your report. The maintainer should be able to quickly reproduce the bug on using the minimal example that you provide.

2. Mention the software architecture on which the bug occurred. 

> In addition to the above, here are the [bug writing guidelines](https://bugs.r-project.org/bugzilla/page.cgi?id=bug-writing.html) on Bugzilla. The [bug reporting](https://www.r-project.org/bugs.html#writing-a-good-bug-report) documentation in R also the discusses practices to write a good bug report.

Reviewer 3: Yes, this probably the main area where we will end up with too many versions of documentation. In time, this guide might replace the guidance on the Bug Reporting page of the R project site, but it is fine to cross-link for now while we are developing the guide.
